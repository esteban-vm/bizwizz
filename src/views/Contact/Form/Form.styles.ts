import clsx from 'clsx'

const styles = {
  groups: clsx('my-6 flex justify-between'),
  half_group: clsx('basis-[48%]'),
  group: clsx('my-6 w-full'),
  feedback: clsx('mt-0 text-darkgrey-2'),
  input: clsx('form-field'),
  textarea: clsx('form-field h-80 resize-none'),
  toast: clsx('text-primary [&_strong]:capitalize'),
}

export default styles
