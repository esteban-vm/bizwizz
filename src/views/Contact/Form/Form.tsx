'use client'
import type { FC } from '@/types'
import { useForm, type SubmitHandler } from 'react-hook-form'
import cogoToast from 'cogo-toast'
import { zodResolver } from '@hookform/resolvers/zod'
import { z } from 'zod'
import { Button } from '@/components'
import styles from './Form.styles'

const validationSchema = z.object({
  name: z.string().trim().min(5, 'Name must be at least 5 characters long'),
  email: z.string().trim().email('Email must be valid'),
  subject: z.string().trim().min(5, 'Subject must be at least 5 characters long'),
  message: z.string().trim().min(10, 'Message must be at least 10 characters long'),
})

type FormValues = z.infer<typeof validationSchema>

const Form: FC = () => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<FormValues>({
    resolver: zodResolver(validationSchema),
    defaultValues: { name: '', email: '', subject: '', message: '' },
    resetOptions: { keepTouched: false, keepDirty: false, keepErrors: false },
  })

  const onSubmit: SubmitHandler<FormValues> = async (data) => {
    reset()

    await cogoToast.success(
      <div className={styles.toast}>
        {Object.entries(data).map(([key, value], index) => (
          <p key={index}>
            <strong>{key}:</strong> {value}
          </p>
        ))}
      </div>,
      { heading: 'Thank you for contacting us!', hideAfter: 5, role: 'alert' }
    )
  }

  return (
    <>
      <h3 id='contact_form_heading'>Shoot Us a Message</h3>
      <div>
        <form aria-labelledby='contact_form_heading' onSubmit={handleSubmit(onSubmit)} noValidate>
          <div className={styles.groups}>
            <div className={styles.half_group}>
              <input
                autoCorrect='off'
                className={styles.input}
                id='name'
                maxLength={30}
                minLength={5}
                placeholder='Your name'
                spellCheck={false}
                type='text'
                {...register('name')}
              />
              {errors.name && <small className={styles.feedback}>{errors.name.message}</small>}
            </div>
            <div className={styles.half_group}>
              <input
                className={styles.input}
                id='email'
                placeholder='Your email address'
                type='email'
                {...register('email')}
              />
              {errors.email && <small className={styles.feedback}>{errors.email.message}</small>}
            </div>
          </div>
          <div className={styles.group}>
            <input
              className={styles.input}
              id='subject'
              maxLength={100}
              minLength={5}
              placeholder='Subject'
              type='text'
              {...register('subject')}
            />
            {errors.subject && <small className={styles.feedback}>{errors.subject.message}</small>}
          </div>
          <div className={styles.group}>
            <textarea
              className={styles.textarea}
              id='message'
              maxLength={500}
              minLength={10}
              placeholder='Your message'
              {...register('message')}
            />
            {errors.message && <small className={styles.feedback}>{errors.message.message}</small>}
          </div>
          <Button text='Send message' isSubmit />
        </form>
      </div>
    </>
  )
}

export default Form
