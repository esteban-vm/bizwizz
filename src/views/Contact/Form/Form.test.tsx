import { render, cleanup, within, screen, userEvent } from '@/tests'
import Form from './Form'

describe('Form:', () => {
  let asFragment: () => DocumentFragment

  beforeEach(() => {
    void ({ asFragment } = render(<Form />))
  })

  afterEach(cleanup)

  it('should render correctly', () => {
    expect(asFragment()).toMatchSnapshot()
  })

  it('should have correct structure', () => {
    const title = /shoot us a message/i
    const formElem = screen.getByRole('form')
    expect(formElem).toBeInTheDocument()
    expect(formElem).toHaveAccessibleName(title)
    expect(within(formElem).getAllByRole('textbox')).toHaveLength(4)
    expect(within(formElem).getByRole('button', { name: /send message/i })).toBeInTheDocument()
  })

  it('should validate input fields', async () => {
    const formElem = screen.getByRole('form')
    const [nameField, emailField, subjectField, messageField] = within(formElem).getAllByRole('textbox')
    const mockValue = 'test'
    await userEvent.type(nameField, mockValue)
    await userEvent.type(emailField, mockValue)
    await userEvent.type(subjectField, mockValue)
    await userEvent.type(messageField, mockValue)
    await userEvent.click(within(formElem).getByRole('button', { name: /send message/i }))
    expect(await within(formElem).findByText(/name must be at least 5 characters long/i)).toBeInTheDocument()
    expect(await within(formElem).findByText(/email must be valid/i)).toBeInTheDocument()
    expect(await within(formElem).findByText(/subject must be at least 5 characters long/i)).toBeInTheDocument()
    expect(await within(formElem).findByText(/message must be at least 10 characters long/i)).toBeInTheDocument()
  })

  it('should be submitted successfully', async () => {
    const formElem = screen.getByRole('form')
    const [nameField, emailField, subjectField, messageField] = within(formElem).getAllByRole('textbox')
    const mockName = 'John Doe'
    const mockEmail = 'test@example.com'
    const mockSubject = 'testing'
    const mockMessage = 'hello world'
    await userEvent.type(nameField, mockName)
    await userEvent.type(emailField, mockEmail)
    await userEvent.type(subjectField, mockSubject)
    await userEvent.type(messageField, mockMessage)
    await userEvent.click(within(formElem).getByRole('button', { name: /send message/i }))
    const alert = await screen.findByRole('alert')
    expect(alert).toBeInTheDocument()
    expect(within(alert).getByRole('heading', { name: /thank you for contacting us!/i, level: 4 })).toBeInTheDocument()
    expect(within(alert).getByText((_, elem) => elem?.textContent === `name: ${mockName}`)).toBeInTheDocument()
    expect(within(alert).getByText((_, elem) => elem?.textContent === `email: ${mockEmail}`)).toBeInTheDocument()
    expect(within(alert).getByText((_, elem) => elem?.textContent === `subject: ${mockSubject}`)).toBeInTheDocument()
    expect(within(alert).getByText((_, elem) => elem?.textContent === `message: ${mockMessage}`)).toBeInTheDocument()
  })
})
