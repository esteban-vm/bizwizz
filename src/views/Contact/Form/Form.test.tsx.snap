// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form: > should render correctly 1`] = `
<DocumentFragment>
  <h3
    id="contact_form_heading"
  >
    Shoot Us a Message
  </h3>
  <div>
    <form
      aria-labelledby="contact_form_heading"
      novalidate=""
    >
      <div
        class="my-6 flex justify-between"
      >
        <div
          class="basis-[48%]"
        >
          <input
            class="form-field"
            maxlength="30"
            minlength="5"
            name="name"
            placeholder="Your name"
            type="text"
          />
        </div>
        <div
          class="basis-[48%]"
        >
          <input
            class="form-field"
            name="email"
            placeholder="Your email address"
            type="email"
          />
        </div>
      </div>
      <div
        class="my-6 w-full"
      >
        <input
          class="form-field"
          maxlength="100"
          minlength="5"
          name="subject"
          placeholder="Subject"
          type="text"
        />
      </div>
      <div
        class="my-6 w-full"
      >
        <textarea
          class="form-field h-80 resize-none"
          maxlength="500"
          minlength="10"
          name="message"
          placeholder="Your message"
        />
      </div>
      <button
        class="block min-w-[20rem] cursor-pointer select-none bg-gradient-120 from-transparent from-0% via-transparent via-50% to-50% bg-[length:250%] p-[1.6rem] text-2xl font-[500] uppercase transition-all duration-300 hover:translate-x-4 hover:bg-[100%] rounded-xl border border-solid border-primary bg-primary to-white text-white hover:text-primary"
        type="submit"
      >
        Send message
      </button>
    </form>
  </div>
</DocumentFragment>
`;
